### DISCUSSION TITLE
Mark destinations as unhealthy by default

### DISCUSSION THREAD

**Discussion by PieterBoeren** [dotnet/yarp#1144] [Q&A] [answered]
Currently, destinations are marked healthy by default when adding a health check (active). Is it possible to mark them as not-healthy?

In my config I set an interval of 10 seconds for the health check, but if the destination is unhealthy, you only get to know this after that interval (and in between, all requests are routed to that destination anyway). I rather have an option that says "the destination is unhealthy until it says otherwise". Is that possible?

**Comment by Tratcher** [comment, marked as answer]
Destinations currently default to `Unknown`, but unknown destinations are still considered when routing requests.

Active health checks are triggered immediately upon server startup, you shouldn't have to wait for the 10s interval. However, you're right that there's a time window at startup where requests could be sent to destinations in an unknown state.

There are a few different ways to handling this, but let's start with one question: Since all destinations are in an unknown state at startup until the first active heath checks complete, how do you want to handle incoming requests? Should they be refused with a 503? Should they be paused until the first health checks complete?

Consider this example:
https://github.com/microsoft/reverse-proxy/blob/82bf8ae27ceb466343163980533b63ac50c938e0/testassets/ReverseProxy.Code/Controllers/HealthController.cs#L20-L34
IActiveHealthCheckMonitor.InitialProbeCompleted reports if the first health check probe has completed. The CheckHealth endpoint is used by a orchestrator to decide if this proxy instance is ready to have requests sent to it. That way destinations are always in a known state before requests are processed.

**Reply by Tratcher** [reply]
The other mechanic available is the IAvailableDestinationsPolicy, but we'll get back to that.

**Reply by PieterBoeren** [reply]
I had the feeling today that the health endpoints weren't triggered on proxy startup, but only after the interval. I'll recheck tomorrow.

I think for me it would come down to treating unknown and unhealthy the same instead of unknown and healthy, thus returning a 503.

**Reply by Tratcher** [reply]
Ok, you can do that by implementing IAvailableDestinationsPolicy.
https://microsoft.github.io/reverse-proxy/articles/dests-health-checks.html#available-destination-collection
You can copy and tweak the default implementation:
https://github.com/microsoft/reverse-proxy/blob/d31b3886d3c5150ed46b2c1ffb4813e6c77b0588/src/ReverseProxy/Health/HealthyAndUnknownDestinationsPolicy.cs#L32

Note that for passive health checks unknown needs to be treated as health so that a destination will be attempted, otherwise you'll never be able to test a destination with live traffic.
